ouroboros-consensus/src/Ouroboros/Consensus/HeaderValidation.hs

```diff
@@ -172,98 +191,145 @@ tickHeaderState :: ConsensusProtocol (BlockProtocol blk)
+    . headerStateHistoryCurrent
+
+-- | Append a 'HeaderState' to the history.
+headerStateHistoryPush ::
+     forall blk.
+     HeaderState blk
+  -> HeaderStateHistory blk
+  -> HeaderStateHistory blk
+headerStateHistoryPush headerState HeaderStateHistory {..} =
+    HeaderStateHistory {
+        headerStateHistorySnapshots = headerStateHistorySnapshots :|> headerState
+      , headerStateHistoryAnchor    = headerStateHistoryAnchor
+      }
 
-castHeaderState :: ( Coercible (ChainDepState (BlockProtocol blk ))
-                               (ChainDepState (BlockProtocol blk'))
-                   , TipInfo blk ~ TipInfo blk'
-                   )
-                => HeaderState blk -> HeaderState blk'
-castHeaderState HeaderState{..} = HeaderState{
-      headerStateConsensus = coerce headerStateConsensus
-    , headerStateTips      = castSeq castAnnTip $ headerStateTips
-    , headerStateAnchor    = fmap    castAnnTip $ headerStateAnchor
+-- | Trim the 'HeaderStateHistory' to the given size, dropping the oldest
+-- snapshots. The anchor will be shifted accordingly.
+--
+-- Note that we do not include the anchor in the size. For example, trimming to
+-- 0 results in no snapshots but still an anchor. Trimming to 1 results in 1
+-- snapshot and an anchor.
+headerStateHistoryTrim ::
+     forall blk. Int
+  -> HeaderStateHistory blk
+  -> HeaderStateHistory blk
+headerStateHistoryTrim n history@HeaderStateHistory {..}
+    | toDrop <= 0
+    = history
+    | otherwise
+    = case Seq.splitAt toDrop headerStateHistorySnapshots of
+        (_ :|> newAnchor, trimmed) -> HeaderStateHistory {
+            headerStateHistorySnapshots = trimmed
+          , headerStateHistoryAnchor    = newAnchor
+          }
+        (Empty, _) ->
+          error $ "impossible: nothing dropped while toDrop = " <> show toDrop
+  where
+    toDrop :: Int
+    toDrop = Seq.length headerStateHistorySnapshots - n
+
+castHeaderStateHistory ::
+     ( Coercible (ChainDepState (BlockProtocol blk ))
+                 (ChainDepState (BlockProtocol blk'))
+     , TipInfo blk ~ TipInfo blk'
+     )
+  => HeaderStateHistory blk -> HeaderStateHistory blk'
+castHeaderStateHistory HeaderStateHistory {..} = HeaderStateHistory {
+      headerStateHistorySnapshots = castSeq castHeaderState headerStateHistorySnapshots
+    , headerStateHistoryAnchor    =         castHeaderState headerStateHistoryAnchor
     }
   where
     -- This is unfortunate. We're doing busy-work on a strict-sequence,
     -- mapping a function that actually doesn't change anything :/
     castSeq :: (a -> b) -> StrictSeq a -> StrictSeq b
-    castSeq f = Seq.fromList . map f . toList
+    castSeq f = Seq.toStrict . fmap f . Seq.getSeq
 
--- | Rewind the header state
+-- | Rewind the header state history
 --
--- This involves 'rewindChainState', and so inherits its PRECONDITION that the
--- target point must have been previously applied.
+-- PRECONDITION: the point to rewind to must correspond to a header (or
```

Comment:
Does an EBB count as a "header" here? I think so, since this is used in ChainSync client.

I would vote for explicitly stating whether EBBs are relevant to "header" or "block" just about everywhere the words are used in a comment, sadly.

Reply:
Done.